File: GroundStation/Libraries/XBeeComm/IMUMessage.cs
using System;
using System.Windows.Media.Media3D;

namespace Baerocats.XBee
{

    public class IMUMessage : Message
    {
        internal const short IMU_MESSAGE_ID = 1;

        Vector3D _acceleration;
        Quaternion _orientation;

        public Vector3D Acceleration
        {
            get
            {
                return _acceleration;
            }
        }

        public Quaternion Orientation
        {
            get
            {
                return _orientation;
            }
        }

        public IMUMessage(uint timestamp, Vector3D acceleration, Quaternion ortientation)
            : base(IMU_MESSAGE_ID, timestamp)
        {
            _acceleration = acceleration;
            _orientation = ortientation;
        }

        public IMUMessage(byte[] data)
            : base(data)
        {
            int readOffset = base.GetDataLength();

            _acceleration = new Vector3D(
                BitConverter.ToSingle(data, readOffset),
                BitConverter.ToSingle(data, readOffset + 4),
                BitConverter.ToSingle(data, readOffset + 8)
                );

            _orientation = new Quaternion(
                BitConverter.ToSingle(data, readOffset + 12),
                BitConverter.ToSingle(data, readOffset + 16),
                BitConverter.ToSingle(data, readOffset + 20),
                BitConverter.ToSingle(data, readOffset + 24)
                );
        }

        internal override int GetDataLength()
        {
            return 28 + base.GetDataLength();
        }

        internal override void GetData(byte[] buffer)
        {
            // The base message fills the buffer first.
            base.GetData(buffer);

            int insertOffset = base.GetDataLength();
            
            Array.Copy(BitConverter.GetBytes(Convert.ToSingle(_acceleration.X)), 0, buffer, insertOffset, 4);
            Array.Copy(BitConverter.GetBytes(Convert.ToSingle(_acceleration.Y)), 0, buffer, insertOffset + 4, 4);
            Array.Copy(BitConverter.GetBytes(Convert.ToSingle(_acceleration.Z)), 0, buffer, insertOffset + 8, 4);
            Array.Copy(BitConverter.GetBytes(Convert.ToSingle(_orientation.W)), 0, buffer, insertOffset + 12, 4);
            Array.Copy(BitConverter.GetBytes(Convert.ToSingle(_orientation.X)), 0, buffer, insertOffset + 16, 4);
            Array.Copy(BitConverter.GetBytes(Convert.ToSingle(_orientation.Y)), 0, buffer, insertOffset + 20, 4);
            Array.Copy(BitConverter.GetBytes(Convert.ToSingle(_orientation.Z)), 0, buffer, insertOffset + 24, 4);
        }
    }
}

File: Arduino/Tracking/IMUMessage.cpp
#include "IMUMessage.h"


IMUMessage::IMUMessage(uint32_t timestamp, imu::Vector<3> acceleration, imu::Quaternion orientation) :
	Message(IMU_MESSAGE_ID, timestamp)
{
	_accelx = acceleration.x();
	_accely = acceleration.y();
	_accelz = acceleration.z();
	_orientationw = orientation.w();
	_orientationx = orientation.x();
	_orientationy = orientation.y();
	_orientationz = orientation.z();
}

IMUMessage::~IMUMessage()
{
}

int IMUMessage::getDataLength() const
{
	return 28 + Message::getDataLength();
}

void IMUMessage::getData(byte * buffer) const
{
	// The base message fills the buffer first.
	Message::getData(buffer);

	byte * insertPtr = buffer + Message::getDataLength();

	memcpy(insertPtr, &_accelx, 4);
	memcpy(insertPtr + 4, &_accely, 4);
	memcpy(insertPtr + 8, &_accelz, 4);
	memcpy(insertPtr + 12, &_orientationw, 4);
	memcpy(insertPtr + 16, &_orientationx, 4);
	memcpy(insertPtr + 20, &_orientationy, 4);  
	memcpy(insertPtr + 24, &_orientationz, 4);
}
